ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -629,6 +646,165 @@ private IndexValue findKey(StoreKey key, FileSpan fileSpan, EnumSet<IndexEntryTy
     return retCandidate;
   }
 
+  /**
+   * Finds all the {@link IndexValue}s associated with the given {@code key} that matches any of the provided {@code types}
+   * if present in the index with the given {@code fileSpan} and return them in reversed chronological order.
+   * @param key the {@link StoreKey} whose {@link IndexValue} is required.
+   * @param fileSpan {@link FileSpan} which specifies the range within which search should be made.
+   * @param types the types of {@link IndexEntryType} to look for.
+   * @param indexSegments the map of index segment start {@link Offset} to {@link IndexSegment} instances
+   * @return The list of the {@link IndexValue}s for {@code key} conforming to one of the types {@code types}.
+   * @throws StoreException any error.
+   */
+  List<IndexValue> findAllIndexValuesForKey(StoreKey key, FileSpan fileSpan, EnumSet<IndexEntryType> types,
+      ConcurrentSkipListMap<Offset, IndexSegment> indexSegments) throws StoreException {
+    List<IndexValue> result = null;
+    final Timer.Context context = metrics.findTime.time();
+    try {
+      ConcurrentNavigableMap<Offset, IndexSegment> segmentsMapToSearch;
+      if (fileSpan == null) {
+        logger.trace("Searching for " + key + " in the entire index");
+        segmentsMapToSearch = indexSegments.descendingMap();
+      } else {
+        logger.trace(
+            "Searching for " + key + " in index with filespan ranging from " + fileSpan.getStartOffset() + " to "
+                + fileSpan.getEndOffset());
+        segmentsMapToSearch = indexSegments.subMap(indexSegments.floorKey(fileSpan.getStartOffset()), true,
+            indexSegments.floorKey(fileSpan.getEndOffset()), true).descendingMap();
+        metrics.segmentSizeForExists.update(segmentsMapToSearch.size());
+      }
+      int segmentsSearched = 0;
+      for (Map.Entry<Offset, IndexSegment> entry : segmentsMapToSearch.entrySet()) {
+        segmentsSearched++;
+        logger.trace("Index : {} searching index with start offset {}", dataDir, entry.getKey());
+        NavigableSet<IndexValue> values = entry.getValue().find(key);
+        if (values != null) {
+          if (result == null) {
+            result = new LinkedList<>();
+          }
+          Iterator<IndexValue> it = values.descendingIterator();
+          while (it.hasNext()) {
+            IndexValue value = it.next();
+            if ((types.contains(IndexEntryType.DELETE) && value.isDelete()) || (types.contains(IndexEntryType.UNDELETE)
+                && value.isUndelete()) || (types.contains(IndexEntryType.TTL_UPDATE) && !value.isDelete()
+                && value.isTTLUpdate()) || (types.contains(IndexEntryType.PUT) && value.isPut())) {
+              // Add a copy of the value to the result since we return a modifiable list to the caller.
+              IndexValue newValue =
+                  new IndexValue(value.getSize(), value.getOffset(), value.getFlags(), value.getExpiresAtMs(),
```

Comment:
Optional: probably worth adding a constructor for IndexValue that takes in another index value for deep copy purpose. (Pass in many parameters is error-prone)

Reply:
make sense, will add a new constructor.